File: Python/AssignProfiles.py
from random import randint, shuffle, choice
import os
from scipy import stats
import DSSEngine


def house_profiles(engine, network, no_feeders, customers_per_feeder, t_month, t_day, directory):
    house = 0
    buses = extract_customer_buses(network, no_feeders, directory)
    occupants_probability = stats.rv_discrete(name='occupants_probability',
                                              values=([i for i in range(1, 6)], (0.30, 0.35, 0.15, 0.13, 0.07)))
    engine.text.command = "set datapath=%s\\Profiles\\House\\" %directory
    for i in range(no_feeders):
        for y in range(customers_per_feeder[i]):
            house += 1
            loadshape = randint(1, 500)
            occupants = occupants_probability.rvs()

            engine.text.command = "new loadshape.Houseload%s npts=1440 minterval=1.0 csvfile=House%s_%s_%s_%s.txt" % (
                house, t_month, t_day, occupants, loadshape)
            engine.text.command = "new load.House%s %s Phases=1 kV=0.23 kW=10 PF=0.97 Daily=Houseload%s" % (
                house, buses[i][y], house)


def pv_clear(engine, network, no_feeders, customers_per_feeder, t_month, clearness, penetration, directory):
    pene_customers = []
    if type(penetration) is float or type(penetration) is int:
        for i in range(no_feeders):
            pene_customers.append(penetration_per_feeder(customers_per_feeder[i], penetration))
    else:
        try:
            for i in range(no_feeders):
                pene_customers.append(penetration_per_feeder(customers_per_feeder[i], penetration[i]))
        except:
            pass
    buses = extract_customer_buses(network, no_feeders)
    engine.text.command = "set datapath=%s\\Profiles\\PV\\Clearness\\" %directory
    iteration = randint(1, 10)
    house = 1

    for i in range(no_feeders):
        shuffle(buses[i])
        for y in range(pene_customers[i]):
            house += 1
            pvsize = randint(1, 4)
            engine.text.command = "new loadshape.PVload%s npts=1440 minterval=1.0 csvfile=PVclear_%s_%s_%s_%s.txt" % (
                house, t_month, clearness, iteration, pvsize)
            engine.text.command = "new generator.PV%s %s Phases=1 kV=0.23 kW=10 PF=1 Daily=PVload%s" % (
            house, buses[i][y], house)


def pv_iter(engine, network, no_feeders, customers_per_feeder, t_month, penetration, directory):
    pene_customers = []
    if type(penetration) is float or type(penetration) is int:
        for i in range(no_feeders):
            pene_customers.append(penetration_per_feeder(customers_per_feeder[i], penetration))
    else:
        try:
            for i in range(no_feeders):
                pene_customers.append(penetration_per_feeder(customers_per_feeder[i], penetration[i]))
        except:
            pass
    buses = extract_customer_buses(network, no_feeders)
    engine.text.command = "set datapath=%s\\Profiles\\PV\\Iterations\\" %directory
    iteration = randint(1, 100)
    house = 1

    for i in range(no_feeders):
        shuffle(buses[i])
        for y in range(pene_customers[i]):
            house += 1
            pvsize = randint(1, 4)
            pvefficiency = choice([1,2])
            engine.text.command = "new loadshape.PVload%s npts=1440 minterval=1.0 csvfile=PViter_%s_%s_%s_%s.txt" % (
            house, t_month, iteration, pvsize, pvefficiency)
            engine.text.command = "new generator.PV%s %s Phases=1 kV=0.23 kW=10 PF=1 Daily=PVload%s" % (
            house, buses[i][y], house)


def extract_customer_buses(network, no_feeders, directory):
    starting_feeder = 63057169

    whole_network = []
    for i in range(no_feeders):
        feeders = []
        with open("%s\\Networks\\%s\\%s\\Loads.txt" % (
                directory, network, (starting_feeder + i))) as f:
            for line in f:
                feeders.append(line.split(" ")[3])

        whole_network.append(feeders)
    return whole_network


def penetration_per_feeder(no_customers, penetration):
    pene_double = penetration * float(no_customers)
    pene_int = int(penetration * no_customers)
    pene_percentage = float(pene_double - pene_int)

    if pene_percentage == 0:
        return pene_int
    else:
        probab = stats.rv_discrete(name='probab',
                                   values=([pene_int, (pene_int + 1)], ((1 - pene_percentage), pene_percentage)))
        return probab.rvs()

File: Matlab/assign_profiles.m
classdef assign_profiles
    properties
        bus;
        customers_per_feeder;
        pv_data;
    end
    methods
        function obj = assign_profiles
            global d;
            obj.customers_per_feeder = d('customers_per_feeder');
            buss = obj.extract_bus_info;
            for i = 1:d('no_feeders');
                tempA = buss(i,1:obj.customers_per_feeder(i));
                idx = randperm(length(tempA));
                tempB(i,1:obj.customers_per_feeder(i))= tempA(idx);
                obj.bus = tempB;
            end
            obj.pv_data = zeros(d('no_customers'),4);
        end
        
        function house(obj)
            global d;
            DSSText = d('DSSText');
            DSSObj = d('DSSObj');
            DSSCircuit = DSSObj.ActiveCircuit;
            
            house = 0;
            %bus = obj.extract_bus_info();
            px = [1 2 3 4 5];
            p = [0.3 0.35 0.15 0.13 0.07];      
            house_data = zeros(d('no_customers'),2);
            
            %Import the house profiles
            house_profiles = matfile([d('mydir'), '\Input\Profiles\House_profiles.mat']);
            temp_house = house_profiles.house(d('month'),d('t_day'),:,:,:);
            %Power Factor
            pf = 0.97;
            reactive_power_mult = tan(acos(pf));
            
            
            for i = 1:d('no_feeders')
                for y = 1:obj.customers_per_feeder(i)
                    
                    house = house + 1;
                    loadshape = randi(500);
                    occupants = randsample(px,1,true,p);
                    
                    house_data(house,1) = occupants;
                    house_data(house,2) = loadshape;
                    
                    temp_array = squeeze(temp_house(1,1,occupants,loadshape,:));
                    DSSCircuit.Loadshapes.New(sprintf('Houseload%u', house));
                    DSSCircuit.Loadshapes.name=sprintf('Houseload%u', house);
                    DSSCircuit.Loadshapes.npts=1440;
                    DSSCircuit.Loadshapes.MinInterval=1;
                    DSSCircuit.Loadshapes.UseActual=1;
                    feature('COM_SafeArraySingleDim',1);
                    DSSCircuit.Loadshapes.Pmult=temp_array;
                    DSSCircuit.Loadshapes.Qmult=(temp_array*reactive_power_mult);
                    feature('COM_SafeArraySingleDim',0);              
                                                          
                    DSSText.Command = sprintf('new load.House%u %s Phases=1 kV=0.23 kW=1 PF=%u Daily=Houseload%u',...
                        house, obj.bus{i,y},pf, house);
                    
                end
            end
            clear temp_house;
            d('house_data') = house_data;
        end
        
        function pv(obj)
            global d;
            DSSText = d('DSSText');
            DSSObj = d('DSSObj');
            DSSCircuit = DSSObj.ActiveCircuit;
            for i = 1:d('no_feeders')
                pv_per_feeder(i) = obj.penetration_per_feeder(obj.customers_per_feeder(i));
            end

            iteration = randi(100);
            %iteration = 2; %CHANGE THIS LATER, FOR DEBUGGING REASONS
            house = 0;          
            
            pv_profiles = matfile([d('mydir'), '\Input\Profiles\PV_profiles.mat']);
            temp_pv = pv_profiles.PV(d('month'), :, :, :, :);
            
            for i = 1:d('no_feeders')
                
                for y = 1:pv_per_feeder(i)
                    house = house + 1;
                    pvsize = randi(4);
                    pvefficiency = randi(2);
                    
                    obj.pv_data(house,1) = iteration;
                    obj.pv_data(house,2) = pvsize;
                    obj.pv_data(house,3) = pvefficiency;
                    
                    temp_array = squeeze(temp_pv(1,iteration,pvsize,pvefficiency,:));
                    DSSCircuit.Loadshapes.New(sprintf('PVload%u', house));
                    DSSCircuit.Loadshapes.name=sprintf('PVload%u', house);
                    DSSCircuit.Loadshapes.npts=1440;
                    DSSCircuit.Loadshapes.MinInterval=1;
                    DSSCircuit.Loadshapes.UseActual=0;
                    feature('COM_SafeArraySingleDim',1);
                    DSSCircuit.Loadshapes.Pmult=temp_array;
                    feature('COM_SafeArraySingleDim',0);
                    
                    
                    DSSCircuit.Loadshapes.New(sprintf('PVStore%u', house));
                    DSSCircuit.Loadshapes.name=sprintf('PVStore%u', house);
                    DSSCircuit.Loadshapes.npts=1440;
                    DSSCircuit.Loadshapes.MinInterval=1;
                    DSSCircuit.Loadshapes.UseActual=0;
                    feature('COM_SafeArraySingleDim',1);
                    DSSCircuit.Loadshapes.Pmult=temp_array*(-1);
                    feature('COM_SafeArraySingleDim',0);
                    
                    DSSText.Command = sprintf('new generator.PV%u %s Phases=1 kV=0.23 kW =1 PF=1 Daily=PVload%u',...
                        house, obj.bus{i,y}, house);
                    
                    if 1==1 %Storage Call
                        obj.storage(house,obj.bus{i,y});
                        obj.pv_data(house,4) = 1;
                    end
                end
                house = sum(obj.customers_per_feeder(1:i));
            end
            clear temp_pv;
            d('pv_data') = obj.pv_data;
            
        end
        
        function storage(obj, house, bus)
            
            global d;
            DSSText = d('DSSText');
            DSSObj = d('DSSObj');
            
            DSSText.Command = sprintf('new storage.battery%u phases=1 %s kV=0.23 kW=1 kWrated=1 kWhrated=7 %%Charge=1 %%stored=0 dispmode=follow daily=PVStore%u',...
                house, bus, house);
            
            
                        
            %DSSText.Command = sprintf('new storage.battery%u phases=1 %s kV=0.23 kWrated=3 kWhrated=7 %%stored=100 State=Discharging dischargetrigger=0.5 daily=PVload1',...
               % house, bus);
            
            %kWrated = charging rate
            %kW = discharging rate
            

        
        end
            
        
        function [number] = penetration_per_feeder(obj,customers)
            global d;
            import assign_house_profiles.*
            
            pene_double = d('pv_penetration') * customers;
            pene_int = int8(d('pv_penetration') * customers);
            pene_percentage = pene_double - double(pene_int);
            
            if pene_percentage == 0
                number = pene_int;
            elseif pene_percentage > 0
                number = randsample([pene_int pene_int+1], 1, true,...
                    [1.0 - pene_percentage pene_percentage]);
            else
                number = randsample([pene_int-1 pene_int], 1, true,...
                    [-pene_percentage 1+pene_percentage]);
            end
            
        end
        
        function [buses] = extract_bus_info(obj)
            global d;
            
            for i = 1:d('no_feeders')
                current_feeder = d('starting_feeder') + i - 1;
                mydir = sprintf('%s//Input//Networks//%s//%u//Loads.txt',...
                    d('mydir'), d('network'), current_feeder);
                fid = fopen(mydir);
                tline = fgets(fid);
                y = 1;
                
                while ischar(tline)
                    new_line = regexp(tline, ' ', 'split');
                    buses(i,y) = new_line(4);
                    y = y +1;
                    tline = fgets(fid);
                end
                fclose(fid);
            end
            
        end
    end
end

